src/ast.ts

```diff
@@ -1099,6 +1103,15 @@ export abstract class TypeNode extends Node {
     }
     return false;
   }
+
+  abstract equals(node: TypeNode): bool;
+
+  static arrayEquals(node: Comparable[] | null, other: Comparable[] | null): bool {
```

Comment:
```ts
static arrayEquals(node: Comparable[] | null, other: Comparable[] | null): bool {
  if (node == other) return true;
  if (!node || !other) return false;
  const len = other.length;
  if (node.length !== len) return false;
  for (let i = len; i-- !== 0;) {
    if (!node[i].equals(other[i])) return false;
  }
  return true;
}
```

Reply:
`node.length !== len` is that needed for number values?